README.md
# pdfmakr

A dart wrapper for PDFMake in browser - [example](http://rxlabz.github.io/pdfmakr)

## Usage

**HTML**

add pdfmake scripts file to your repo and your html page

```html
<script src='js/pdfmake.min.js'></script>
<script src='js/vfs_fonts.js'></script>
```

**Dart**

```dart
Future downloadPdf() async {
  final pdfFile = await pdf.createPdf(new pdf.PDFContent(
      content: [
        new pdf.Image(image: await loadImg64("img/rx.png")),
        new pdf.TextBloc(text: "text example", bold: true, color: 'blue'),
        new pdf.TextBloc(text: "text example with style", style: 'header'),
        new pdf.TextBloc(text: " "),
        new pdf.Columns(columns: [
          new pdf.Column(text: 'col 1 80%', width: '80%'),
          new pdf.Column(text: 'col 2'),
        ]),
        new pdf.TextBloc(text: " "),
        new pdf.Table(
            table: new pdf.TableBody(body: [
          [
            new pdf.TextBloc(text: "cell example 1", bold: true, color: 'blue'),
            new pdf.TextBloc(
                text: "cell example - col 2",
                bold: true,
                color: 'green',
                fillColor: 'yellow'),
          ],
          [
            new pdf.TextBloc(text: "bold row 2 - col 1", bold: true),
            new pdf.TextBloc(text: "italic row 2 - col 2", italics: true),
          ]
        ]))
      ],
      styles: new pdf.Styles(
          header: new pdf.Style(color: 'red', fontSize: 24),
          basic: new pdf.Style(color: 'green'))));

  pdfFile.download();
  //pdfFile.print();
  //pdfFile.open();
}

Future<String> loadImg64(String url) async {
  HttpRequest response =
      await HttpRequest.request(url, responseType: "arraybuffer");

  String contentType = response.getResponseHeader('Content-Type');

  var list = new Uint8List.view((response.response as ByteBuffer));

  String header = 'data:$contentType;base64,';
  String base64 = BASE64.encode(list);
  String image = "${header}${base64}";

  return image;
}

```

## Example

works with DDC

```bash
cd example
pub get
pub serve

# open in chrome
```

## Features / Todos

- [x] createPdf
- Content types
  - [x] text
  - [ ] image :
    - [x] url
    - [ ] size
  - [ ] Columns
    - [x] text, width
    - [ ] multitype columns content
  - [ ] Table
    - [x] body
    - [ ] more ?
- [x] Styles dictionaries ( user needs implements its own class )
- [x] pdf.download()
- [x] pdf.open()
- [x] pdf.print()

- dart-ier api : cascade ?

- lists : ul, ol
- headers, footers
- background layer
- margins
- text decorations
- paragraphs
- image sizes
- pageBreak
- page dimensions, orientation and margins
- more internal properties
- browser preview
 */


CHANGELOG.md
# Changelog

## 0.0.2
- Add support for Dart 2.0
  - example updated to use webdev for building and serving of content

## 0.0.1

- createPdf(PDFContent)
- Content types
  - text : size, color, bold, italic
  - Image : url
  - basic text Columns
  - Table : body
- Styles dictionaries ( implies implementing its own class )
- pdfDoc.download(),pdfDoc.open(),pdfDoc.print()


lib/src/pdf.dart
@JS('pdfMake')
library pdf;

import 'package:js/js.dart';


@JS("createPdf")
external Doc create(PDFContent content);


@JS()
@anonymous
class PDFContent {
  external List<dynamic> get content;
  external Styles get styles;
  external factory PDFContent({List<dynamic> content, Styles styles});
}

@JS()
@anonymous
class Doc {
  external void download();
  external void print();
  external void open();
  external factory Doc();
}

@JS()
@anonymous
class TextBloc {
  external String get text;
  external String get fillColor;
  external String get fontSize;
  external bool get bold;
  external bool get italics;
  external bool get color;
  external String get style;
  external factory TextBloc(
    {String text,
      String style,
      String fillColor = 'white',
      int fontSize = 12,
      bool bold = false,
      bool italics = false,
      String color});
}

@JS()
@anonymous
class Styles {
  external Style get header;
  external Style get basic;
  external factory Styles({Style header, Style basic});
}

@JS()
@anonymous
class Image {
  external String get image;
  external factory Image({String image});
}

@JS()
@anonymous
class Columns {
  external List<Column> get columns;
  external int get columnGap;

  external factory Columns({List<Column> columns, int columnGap = 0});
}

@JS()
@anonymous
class Column {
  external String get width;
  external String get text;
  external String get fillColor;
  external factory Column(
      {String text, String width, String fillColor = 'yellow'});
}

@JS()
@anonymous
class Table {
  external TableBody get table;
  external factory Table({TableBody table});
}

@JS()
@anonymous
class TableBody {
  external List<dynamic> get body;
  external factory TableBody({List<dynamic> body});
}

@JS()
@anonymous
class Style {
  external String get fontSize;
  external bool get bold;
  external bool get italic;
  external bool get color;
  external factory Style({
    String color = 'black',
    int fontSize = 12,
    bool bold = false,
    bool italic = false,
  });
}



example/web/main.dart
import 'dart:async';
import 'dart:html';
import 'dart:typed_data';
import 'package:dart2_constant/convert.dart' as convert;
import 'package:pdfmakr/pdfmakr.dart' as pdf;

String img64;

void main() {
  querySelector('#btPdf').onClick.listen((e) => downloadPdf());
}

Future<String> loadImg64(String url) async {
  HttpRequest response =
      await HttpRequest.request(url, responseType: "arraybuffer");

  String contentType = response.getResponseHeader('Content-Type');

  var list = new Uint8List.view((response.response as ByteBuffer));

  String header = 'data:$contentType;base64,';
  String base64 = convert.base64.encode(list);
  String image = "${header}${base64}";

  return image;
}

Future downloadPdf() async {
  final pdfFile = await pdf.create(new pdf.PDFContent(
      content: [
        new pdf.Image(image: await loadImg64("img/rx.png")),
        new pdf.TextBloc(text: "text example", bold: true, color: 'blue'),
        new pdf.TextBloc(text: "text example with style", style: 'header'),
        new pdf.TextBloc(text: " "),
        new pdf.Columns(columns: [
          new pdf.Column(text: 'col 1 80%', width: '80%'),
          new pdf.Column(text: 'col 2'),
        ]),
        new pdf.TextBloc(text: " "),
        new pdf.Table(
            table: new pdf.TableBody(body: [
          [
            new pdf.TextBloc(text: "cell example 1", bold: true, color: 'blue'),
            new pdf.TextBloc(
                text: "cell example - col 2",
                bold: true,
                color: 'green',
                fillColor: 'yellow'),
          ],
          [
            new pdf.TextBloc(text: "bold row 2 - col 1", bold: true),
            new pdf.TextBloc(text: "italic row 2 - col 2", italics: true),
          ]
        ]))
      ],
      styles: new pdf.Styles(
          header: new pdf.Style(color: 'red', fontSize: 24),
          basic: new pdf.Style(color: 'green'))));

  pdfFile.download();
}
